### PR TITLE
Java frontend: create String and Class literals earlier

### PR DESCRIPTION
**Pull request by smowton**
This creates String and Class literals ahead of running java_bytecode_convert_method
instead of creating them in an after-the-fact fixup in typecheck_expr as previously.
These now exist before method conversion begins and __CPROVER_initialize is created,
which means the String literals are correctly initialised even when using incremental
method conversion.

This can be a little wasteful if they are only referred to from methods that would never
truly be invoked, but preliminary experiments suggest that creating ~10000 Strings takes
under a second, and that Java lazy loading can quickly delete ones that are not in fact
used.

This "right first time" approach to Java conversion has similar benefits to https://github.com/diffblue/cbmc/pull/1764, and makes incremental function loading easier.

### CODE REVIEW

File: src/java_bytecode/java_bytecode_language.cpp [hunk 1]
```diff
@@ -195,6 +196,87 @@ bool java_bytecode_languaget::parse(
...
+          address_of_exprt address_of_expr(symbol_expr);
+          ldc_arg0 = address_of_expr;
+
+          if(!symbol_table.has_symbol(symbol_expr.get_identifier()))
+          {
+            symbolt new_class_symbol;
+            new_class_symbol.name = symbol_expr.get_identifier();
+            new_class_symbol.type = symbol_expr.type();
+            // name without "java::" prefix:
+            new_class_symbol.base_name =
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_language.cpp]:
```C++
INVARIANT(has_suffix(new_class_symbol.name, "java::")
```

File: src/java_bytecode/java_bytecode_language.cpp [hunk 2]
```diff
@@ -195,6 +196,87 @@ bool java_bytecode_languaget::parse(
...
+        // onto the stack.
+
+        // 3) Pushing a Class constant causes a reference to a java.lang.Class
+        // to be pushed onto the operand stack
+
+        exprt &ldc_arg0 = instruction.args[0];
+
+        if(ldc_arg0.id() == ID_type)
+        {
+          irep_idt class_id = ldc_arg0.type().get(ID_identifier);
```
Comment (thk123) [inline src/java_bytecode/java_bytecode_language.cpp]: I'd be tempted to pull this into a `get_or_create_class_constant` or something

File: src/java_bytecode/java_string_literals.cpp [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+#include <linking/zero_initializer.h>
+
+#include <util/arith_tools.h>
+#include <util/namespace.h>
+#include <util/unicode.h>
+
+#include <iomanip>
+#include <sstream>
+
+static std::string escape_non_alnum(const std::string &toescape)
```
Comment (thk123) [inline src/java_bytecode/java_string_literals.cpp]: I appreciate you are just moving a function, but now would be a great time to add some doxygen to this function (I for one have no idea what an alnum is and why I'd want to escape it). 
Nit: `toescape` -> `to_escape`

**Comment by smowton** [comment]
TG bump: https://github.com/diffblue/test-gen/pull/1413

**Review by thk123 — DISMISSED** [review]
TG pointer bump failing on a bunch of string tests...

**Comment by smowton** [comment]
@thk123 changes applied, and I think I've found the cause of the test failures.

@romainbrenguier @allredj it appears when String constant globals are made earlier and string-refinement is in play, then two ancillary globals are created:

1. `Literal.xyz.constarray = { 'x', 'y', 'z' }`. This was being eliminated by ci-lazy-methods' unused-globals sweep, which I fix here by augmenting it slightly to notice that Literal.xyz's initialiser referring to it implies it is needed.

2. `Literal.xyz.return_value = associate_array_to_pointer({'x', 'y', 'z'}, &Literal.xyz)`. This was stranger -- if I'm reading it correctly, this is never used and is just there to force that associate function to "run" (except it is declared as a mathematical function, so it usually isn't supposed to have side-effects). I've worked around this by initialising Literal.xyz with `Literal.xyz.return_value, &Literal.xyz.constarray` (as in a C comma-expression), intending to indicate that "if you need Literal.xyz then you also need Literal.xyz_return_value", which seems to be enough to convince everybody that the return_value expression is needed. 

However this is pretty fragile, since the mathematical function construct is being somewhat abused here since its return value is unimportant, and I would expect a future slightly-cleverer unused-globals analysis to delete it again.

In summary I recommend coming up with a better way to express the array <-> String pointer association that doesn't use a mathematical-function-with-side-effect.

**Comment by smowton** [comment]
Fixed failure, was caused by this patch promoting `@class_model` objects to constant globals (which they always should have been). They contain HashTable members which get a `cproverNondetInitialize` call, which in turn could fail because the object factory could run the init function at the wrong time. I'll file a separate PR for the fix; this PR for the time being depends on that one.

**Comment by smowton** [comment]
The nondetInitialize fix has been split into a separate PR: https://github.com/diffblue/cbmc/pull/1769

File: src/java_bytecode/java_bytecode_language.cpp [hunk 4]
```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
...
+  symbol_exprt symbol_expr(
+    id2string(class_id)+"@class_model",
+    java_lang_Class);
+
+  if(!symbol_table.has_symbol(symbol_expr.get_identifier()))
+  {
+    symbolt new_class_symbol;
+    new_class_symbol.name = symbol_expr.get_identifier();
+    new_class_symbol.type = symbol_expr.type();
+    // name without "java::" prefix:
```
Comment (LAJW) [inline src/java_bytecode/java_bytecode_language.cpp]: I think this comment should be removed.

File: src/java_bytecode/java_bytecode_language.cpp [hunk 5]
```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
...
+  bool string_refinement_enabled)
+{
+  for(auto &method : parse_tree.parsed_class.methods)
+  {
+    for(auto &instruction : method.instructions)
+    {
+      if(instruction.statement == "ldc" ||
+         instruction.statement == "ldc2" ||
+         instruction.statement == "ldc_w" ||
+         instruction.statement == "ldc2_w")
```
Comment (LAJW) [inline src/java_bytecode/java_bytecode_language.cpp]: It'd be great if you could pull this condition into a function. I'm not sure what any of those magic values mean.

File: src/java_bytecode/java_bytecode_language.cpp [hunk 6]
```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
...
+          "ldc instructions should have an argument");
+
+        // 1) Pushing a String causes a reference to a java.lang.String object
+        // to be constructed and pushed onto the operand stack.
+
+        // 2) Pushing an int or a float causes a primitive value to be pushed
+        // onto the stack.
+
+        // 3) Pushing a Class constant causes a reference to a java.lang.Class
+        // to be pushed onto the operand stack
```
Comment (LAJW) [inline src/java_bytecode/java_bytecode_language.cpp]: These comments describe "what" and not why, which renders them unnecessary.

File: src/java_bytecode/java_bytecode_language.cpp [hunk 7]
```diff
@@ -256,6 +257,106 @@ static void infer_opaque_type_fields(
...
+
+        exprt &ldc_arg0 = instruction.args[0];
```
Comment (LAJW) [inline src/java_bytecode/java_bytecode_language.cpp]: I'd mark it `const exprt &ldc_arg0` here, and then assign `instruction.args[0]` to a closure. Would make it clear that this is not modified in called functions, only reassigned.

File: src/java_bytecode/java_bytecode_language.cpp [hunk 8]
```diff
@@ -317,6 +418,23 @@ bool java_bytecode_languaget::typecheck(
     infer_opaque_type_fields(c.second, symbol_table);
   }
 
+  // Create global variables for constants (String and Class literals) up front.
+  // This means that when running with lazy loading, we will be aware of these
+  // literal globals' existence when __CPROVER_initialize is generated in
+  // `generate_support_functions`.
+  std::size_t before_constant_globals_size = symbol_table.symbols.size();
```
Comment (LAJW) [inline src/java_bytecode/java_bytecode_language.cpp]: `const`, please?

File: src/java_bytecode/java_string_literals.cpp [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+      escaped << ch;
+    else
+      escaped << '_'
+              << std::hex
+              << std::setfill('0')
+              << std::setw(2)
+              << (unsigned int)ch;
+  }
+  return escaped.str();
+}
```
Comment (LAJW) [inline src/java_bytecode/java_string_literals.cpp:44]: Moving those functions into a separate file should be done in a different commit. Right now it looks similar, but not identical.

Edit: Commit separated, I commend.

File: src/java_bytecode/java_string_literals.cpp [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+  }
+
+  bool add_failed = symbol_table.add(new_symbol);
+  INVARIANT(
+    !add_failed,
+    "string literal symbol was already checked not to be "
+    "in the symbol table, so adding it should succeed");
+
+  return new_symbol.symbol_expr();
+}
```
Comment (LAJW) [inline src/java_bytecode/java_string_literals.cpp:210]: This function is way too long (but I suppose it would be acceptable to put it here as a move-only commit).

Edit: Fine, as it has been put into a move-only commit.

**Review by LAJW — CHANGES_REQUESTED** [review]
The only thing that kills it for me is the fact, that moving code around isn't done in a dedicated commit.

File: src/java_bytecode/java_bytecode_language.cpp (hunk 4 above)
Comment (smowton) [reply src/java_bytecode/java_bytecode_language.cpp]: Not done; without it it is not clear what we're trying to clip by removing 6 chars.

File: src/java_bytecode/java_bytecode_language.cpp (hunk 5 above)
Comment (smowton) [reply src/java_bytecode/java_bytecode_language.cpp]: Added a comment. No other Java bytecode opcodes have such functions, so I'll leave these ones plain as well for now.

File: src/java_bytecode/java_bytecode_language.cpp (hunk 6 above)
Comment (smowton) [reply src/java_bytecode/java_bytecode_language.cpp]: Not done; this describes the semantics of the opcode more clearly than reading the code.

File: src/java_bytecode/java_bytecode_language.cpp (hunk 7 above)
Comment (smowton) [reply src/java_bytecode/java_bytecode_language.cpp]: Split it into a separate function.

**Comment by smowton** [comment]
@LAJW split the move apart from the substantial commit, and implemented those suggestions not otherwise responded to above.

**Comment by smowton** [comment]
@LAJW please re-review; @peterschrammel @thk123 @tautschnig this is ready for review

**Review by thk123 — APPROVED** [review]
Change looks good, TG pointer bump is passing

File: src/java_bytecode/java_bytecode_language.cpp [hunk 11]
```diff
@@ -256,6 +257,118 @@ static void infer_opaque_type_fields(
...
+        get_or_create_class_literal_symbol(class_id, symbol_table));
+  }
+  else if(ldc_arg0.id() == ID_java_string_literal)
+  {
+    return
+      address_of_exprt(
+        get_or_create_string_literal_symbol(
+          ldc_arg0, symbol_table, string_refinement_enabled));
+  }
+  else
```
Comment (LAJW) [inline src/java_bytecode/java_bytecode_language.cpp:325]: This else is redundant.

**Review by LAJW — APPROVED** [review]
It's fine.
